### DISCUSSION TITLE
Crashes caused by repeated calls of m_get_screen_coords_for_world_coords?

### DISCUSSION THREAD

**Discussion by rkwapisz** [YimMenu/YimMenu#2925] [Q&A]
I'm working on a bone ESP enhancement and have been running into some very strange menu crashes that I can attribute to a peculiarity with m_get_screen_coords_for_world_coords.

I've written a helper function that takes in an ImVec2 pointer and writes the W2S coordinates derived from the player and a defined ePedBoneType. This function is called multiple times from draw_player (view_esp.cpp), and the results are used with ImGui's AddLine to draw the bones on screen.

```
void esp::bone_to_screen(const player_ptr& plyr, ePedBoneType pedBone, ImVec2& boneVec)
{
if (plyr == nullptr)
	return;

float bone_x = 0;
float bone_y = 0;

float screenX = (float)*g_pointers->m_gta.m_resolution_x;
float screenY = (float)*g_pointers->m_gta.m_resolution_y;

g_pointers->m_gta.m_get_screen_coords_for_world_coords(plyr->get_ped()->get_bone_coords(pedBone).data, &bone_x, &bone_y);

boneVec.x = screenX * bone_x;
boneVec.y = screenY * bone_y;
}
```

Then in view_esp.cpp, a sample of what I'm doing with a few bones:
```
ImVec2 r_hand_pos;
bone_to_screen(plyr, ePedBoneType::R_HAND, r_hand_pos);

ImVec2 l_hand_pos;
bone_to_screen(plyr, ePedBoneType::L_HAND, l_hand_pos);

// Neck to hands
draw_list->AddLine(neck_pos, r_hand_pos, esp_color, 2.0f);
draw_list->AddLine(neck_pos, l_hand_pos, esp_color, 2.0f);
```

Calling m_get_screen_coords_for_world_coords has a nasty tendency to crash the game. When I tried debugging, I added a LOG(INFO) statement before the call and, to my surprise, it allowed all the bones to render correctly without crashing.

```
LOG(INFO) << "BEFORE g_pointers->m_gta.m_get_screen_coords_for_world_coords(boneData, &bone_x, &bone_y);";
g_pointers->m_gta.m_get_screen_coords_for_world_coords(plyr->get_ped()->get_bone_coords(pedBone).data, &bone_x, &bone_y);
```

I suspect the I/O delay of LOG(INFO) compensates for some strange race conditions in m_get_screen_coords_for_world_coords, but I can't be too sure. I want to stay as far away from calling natives inside a rendering thread, so I don't know what other options I have for efficiently accessing player coordinates inside the ESP function.

**Comment by xiaoxiao921** [comment]
reimplementing should be easy given we have grcViewport in the gtav class repo now

**Reply by rkwapisz** [reply]
Could be worth checking out. I won't get into profiling quite yet, but I'm curious to see if GET_SCREEN_COORD_FROM_WORLD_COORD is much slower.
